parsing/location.mli

```diff
@@ -26,7 +26,27 @@ type t = Warnings.loc = {
   loc_start: Lexing.position;
   loc_end: Lexing.position;
   loc_ghost: bool;
-}
+  }
+(** [t] represents a segment of source code between two positions.
+
+    For positions coming from a parser, loc_ghost is usually false, which
+    indicates that the information attached to the location could in principle
+    be parsed from the location. This is used by tools like merlin that want to
+    relate source code with parsetrees or later asts.
+
+    Example: in `let f x = x`, we have:
+    - a structure item at location "let f x = x"
+    - a pattern "f" at location "f"
+    - an expression "fun x -> x" at location "x = x"
+    - a pattern "x" at location "x"
+    - an expression "x" at location "x"
+    In this case, every node has loc_ghost=false, except the node "fun x -> x",
+    since [Parser.expression (Lexing.from_string "x = x")] would fail to parse.
+    By contrast, in `let f = fun x -> x`, every node has loc_ghost=false.
+
+    The presence of line directives could create locations that don't follow
+    this condition, but that seems like a theoretical concern.
```

Comment:
Line directives do not influence if a location is a ghost or not?

Reply:
I have not considered this possibility, that would be wild.
Empirically, the answer is no:

```ocaml
#1 "foo"
let
#1 "bar"
x = 1
```

parsetree:

```
[
  structure_item (foo[1,10+0]..bar[1,23+5])
    Pstr_value Nonrec
    [
      <def>
        pattern (bar[1,23+0]..[1,23+1])
          Ppat_var "x" (bar[1,23+0]..[1,23+1])
        expression (bar[1,23+4]..[1,23+5])
          Pexp_constant PConst_int (1,None)
    ]
]
```

Reply:
I tweaked the text a bit. I claim that loc_ghost is about mapping between the ast and the parsed string, and line directives are about mapping between the parsed string and what the user considers the real source code. So essentially line directives should be thought of as orthogonal to line directives.